### PR TITLE
llama: limit max outputs of `llama_context`

### PR DESCRIPTION
**Pull request by am17an**
## Overview

continue #23764, this PR only reserves logits space for `n_seqs` when possible. With `-ub 2048` and MTP, this saves another 1.2GB of VRAM for me. I've tested `llama-perplexity` also and it seems to work fine. ~But maybe there is a better API, putting up as a draft for now~ According to me an API in llama-context is a good solution for this, by default it will reserve all tokens but specifically in server-context we can set it to 1 whenever possible.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES, for review and testing across various configs

### CODE REVIEW

**Comment by drauh** [comment]
To motivate this PR further, here is the experiment I ran (copied from the issue #23527 which would be closed by this PR). This patch provides memory reduction even if not using MTP leaving some headroom to put more experts on the GPU for instance.

Setup: Qwen3.6-35B-A3B-UD-Q4_K_XL, `-c 150000 -b 4096 -ub 4096 -fa on -ctk q8_0 -ctv q8_0 --kv-unified --cpu-moe -np 1`, RTX 3080 10 GB. Peak sampled at 10 Hz over a 145041-token PP.

| run | CUDA0 compute | CUDA_Host compute | idle VRAM | peak VRAM |
|---|---:|---:|---:|---:|
| baseline | 3976 MiB | 2408 MiB | 7981 MiB | 8275 MiB |
| #23861 only | 3892 MiB | 2408 MiB | 7897 MiB | 8195 MiB |
| #23764 only | 3976 MiB | 1236 MiB | 7979 MiB | 8277 MiB |
| #23861+ #23764 | **1982 MiB** | 1236 MiB | **5985 MiB** | **6279 MiB** |

**Comment by joaotbelo** [comment]
Corroborating data on **Blackwell (SM_120, RTX 5070 Ti 16 GB)** for a config not yet in this thread: **MTP draft decoding + vision (mmproj) + `--parallel 2`**, real assistant workload.

Built this branch (`810aa71`, includes #23764) for `CUDA_DOCKER_ARCH=120` and compared against `b9426` (already carries #23764), so this isolates the #23861 increment.

**Config (identical both runs):** `Qwen3.6-35B-A3B-UD-Q2_K_XL-MTP`, `-c 131072 --parallel 2`, `--spec-type draft-mtp`, vision mmproj loaded, `-ctk q4_0 -ctv q4_0`, `-fa on`, default `-b 2048 -ub 512`. Workload: 1 vision request + 2× concurrent ~5.6k-token prefill → 256-tok decode; peak sampled at 5 Hz.

| run | idle VRAM | peak VRAM |
|---|---:|---:|
| b9426 (#23764 only) | 15495 MiB | 15513 MiB |
| **+ #23861** | **14835 MiB** | **14859 MiB** |
| saving | **660 MiB** | **654 MiB** |

Smaller than the `-ub 4096` numbers above (expected — the reservation scales with `n_ubatch`; we run the default `-ub 512`), but on a 16 GB card it nearly doubles free headroom at peak (790 → 1444 MiB). No OOM, MTP draft acceptance unchanged (~0.5–0.72), and the saving holds at idle and under load — so it's reclaiming a persistently-reserved buffer, not just a prefill spike. 👍

**Review by pwilkin — APPROVED** [review]
Let's do it!

**Review by ServeurpersoCom — APPROVED** [review]

File: src/llama-cparams.h [hunk 1]
```diff
@@ -12,9 +12,10 @@ struct llama_cparams {
     uint32_t n_batch;
     uint32_t n_ubatch;
     uint32_t n_seq_max;
-    uint32_t n_rs_seq;        // number of recurrent-state snapshots per seq for rollback
-    int32_t  n_threads;       // number of threads to use for generation
-    int32_t  n_threads_batch; // number of threads to use for batch processing
+    uint32_t n_rs_seq;          // number of recurrent-state snapshots per seq for rollback
+    uint32_t n_outputs_per_seq; // max outputs per sequence in a ubatch (0 = no limit)
```
Comment (ggerganov) [inline src/llama-cparams.h]: This is not correct technically. This information (i.e. "how many outputs") is something that has to be associated with the batch, not with the context. Unfortunately, the current `llama_batch` design does not allow this (along with many other limitations).

Comment (am17an) [reply src/llama-cparams.h]: I see, is there a way to add this to llama-ext?

Comment (ggerganov) [reply src/llama-cparams.h]: - Rename this field to `n_outputs_max`, (default = 0, equal to `n_batch`)
- Move the `graph_n_outputs_pp` logic to the server code. It determines the `cparams.n_outputs_max` and passes it to the context upon creation
- Add asserts in `llama_context` that the outputs in the passed batches do not exceed `cparams.n_outputs_max`

Comment (ggerganov) [reply src/llama-cparams.h]: The idea is that when creating a context, we should be able to set a limit for the max number of outputs that it will support. By default, all tokens in a batch could be outputs - this has always been the case until now. With specifying `n_outputs_max`, we limit the amount, but need some guardrails (i.e. the asserts) when the user tries to process a batch with more outputs. This change is independent of any other logic (technically, it can be separate PR, but it's ok).

Then in the server, we add logic to determine how much outputs we are going to be producing. For example, if the server is going to be used to compute embeddings, then we don't need to change the default `n_outputs_max`. And if the server is going to be used for decoding it needs one output per sequence + any extra stuff from the speculative decoding settings. I.e. you need to combine the `server_n_outputs_per_seq` and `graph_n_outputs_pp` logics and move both in the server, right before creating the contexts.

Comment (am17an) [reply src/llama-cparams.h]: I added the assert and tested a couple of combinations

File: common/common.cpp [hunk 2]
```diff
@@ -1562,7 +1562,8 @@ struct llama_context_params common_context_params_to_llama(const common_params &
 
     cparams.n_ctx             = params.n_ctx;
     cparams.n_seq_max         = params.n_parallel;
-    cparams.n_rs_seq          = params.speculative.need_n_rs_seq();
+    cparams.n_rs_seq      = params.speculative.need_n_rs_seq();
+    cparams.n_outputs_max = std::max(params.n_outputs_max, 0);
```
Comment (ggerganov) [inline common/common.cpp]:
```suggestion
    cparams.n_rs_seq          = params.speculative.need_n_rs_seq();
    cparams.n_outputs_max     = std::max(params.n_outputs_max, 0);
```

File: common/common.h [hunk 3]
```diff
@@ -431,6 +431,7 @@ struct common_params {
     int32_t n_chunks              =    -1; // max number of chunks to process (-1 = unlimited)
     int32_t n_parallel            =     1; // number of parallel sequences to decode
     int32_t n_sequences           =     1; // number of sequences to decode
+    int32_t n_outputs_max         =     0; // max outputs in a ubatch (0 = n_batch)
```
Comment (ggerganov) [inline common/common.h]:
```suggestion
    int32_t n_outputs_max         =     0; // max outputs in a batch (0 = n_batch)
```

File: include/llama.h [hunk 4]
```diff
@@ -339,6 +339,7 @@ extern "C" {
         uint32_t n_ubatch;          // physical maximum batch size
         uint32_t n_seq_max;         // max number of sequences (i.e. distinct states for recurrent models)
         uint32_t n_rs_seq;          // number of recurrent-state snapshots per seq for rollback (0 = no rollback) [EXPERIMENTAL]
+        uint32_t n_outputs_max;     // max outputs in a ubatch (0 = n_batch)
```
Comment (ggerganov) [inline include/llama.h:342]:
```suggestion
        uint32_t n_outputs_max;     // max outputs supported by the context (0 = n_batch)
```

File: src/llama-context.cpp [hunk 5]
```diff
@@ -577,16 +579,18 @@ void llama_context::sched_reserve() {
     int n_splits_tg = -1;
     int n_nodes_tg  = -1;
 
+    const uint32_t n_outputs_pp = std::min(n_tokens, cparams.n_outputs_max);
```
Comment (ggerganov) [inline src/llama-context.cpp:582]:
```suggestion
    const uint32_t n_outputs_max = std::min(n_tokens, cparams.n_outputs_max);
```

File: src/llama-context.cpp [hunk 6]
```diff
@@ -774,7 +778,9 @@ bool llama_context::memory_update(bool optimize) {
         const uint32_t n_seqs = cparams.n_seq_max;
         const uint32_t n_tokens = std::min(cparams.n_ctx, cparams.n_ubatch);
 
-        auto * gf = graph_reserve(n_tokens, n_seqs, n_tokens, mctx.get());
+        const uint32_t n_outputs_pp = std::min(n_tokens, cparams.n_outputs_max);
```
Comment (ggerganov) [inline src/llama-context.cpp]:
```suggestion
        const uint32_t n_outputs_max = std::min(n_tokens, cparams.n_outputs_max);
```

File: src/llama-context.cpp [hunk 7]
```diff
@@ -1780,6 +1786,8 @@ int llama_context::decode(const llama_batch & batch_inp) {
 
             // needs to happen before the graph is built
             n_outputs = n_outputs_new;
+
+            GGML_ASSERT(n_outputs <= cparams.n_outputs_max);
```
Comment (ggerganov) [inline src/llama-context.cpp]: The `n_outputs` is for the current ubatch. We have to check against `n_outputs_all` - for the entire batch.

Best to move the assert to `output_reserve()`

File: src/llama-context.cpp [hunk 8]
```diff
@@ -2224,9 +2232,13 @@ ggml_cgraph * llama_context::graph_reserve(
     LLAMA_LOG_DEBUG("%s: reserving a graph for ubatch with n_tokens = %4u, n_seqs = %2u, n_outputs = %4u\n", __func__, n_tokens, n_seqs, n_outputs);
     GGML_ASSERT(n_outputs >= 1);
 
+    const bool reserve_all_outputs = n_outputs >= n_tokens;
```
Comment (ggerganov) [inline src/llama-context.cpp]: Is this needed?

File: src/llama-cparams.h [hunk 9]
```diff
@@ -13,6 +13,7 @@ struct llama_cparams {
     uint32_t n_ubatch;
     uint32_t n_seq_max;
     uint32_t n_rs_seq;        // number of recurrent-state snapshots per seq for rollback
+    uint32_t n_outputs_max;   // max outputs in a ubatch
```
Comment (ggerganov) [inline src/llama-cparams.h]:
```suggestion
    uint32_t n_outputs_max;   // max outputs supported by the context
```

File: tools/server/server-context.cpp [hunk 10]
```diff
@@ -37,6 +37,50 @@ using json = nlohmann::ordered_json;
...
 constexpr int HTTP_POLLING_SECONDS = 1;
 
+static uint32_t server_n_outputs_max(const common_params & params) {
+    const uint32_t n_batch  = params.n_batch;
+    const uint32_t n_ubatch = std::min(n_batch, params.n_ubatch == 0 ? n_batch : params.n_ubatch);
+
+    if (params.embedding ||
+            (params.pooling_type != LLAMA_POOLING_TYPE_UNSPECIFIED && params.pooling_type != LLAMA_POOLING_TYPE_NONE)) {
+        return n_ubatch;
+    }
```
Comment (ggerganov) [inline tools/server/server-context.cpp:46]:
```suggestion
    const uint32_t n_batch  = params.n_batch;
    
    if (params.embedding ||
            (params.pooling_type != LLAMA_POOLING_TYPE_UNSPECIFIED && params.pooling_type != LLAMA_POOLING_TYPE_NONE)) {
        return n_batch;
    }
```

File: tools/server/server-context.cpp [hunk 11]
```diff
@@ -37,6 +37,50 @@ using json = nlohmann::ordered_json;
...
+                break;
+            case COMMON_SPECULATIVE_TYPE_NONE:
+            case COMMON_SPECULATIVE_TYPE_COUNT:
+                break;
+        }
+    }
+
+    const uint64_t n_outputs = (uint64_t) params.n_parallel * n_outputs_per_seq;
+
+    return std::max<uint32_t>(1, std::min<uint64_t>(n_ubatch, n_outputs));
```
Comment (ggerganov) [inline tools/server/server-context.cpp]:
```suggestion
    return std::max<uint32_t>(1, std::min<uint64_t>(n_batch, n_outputs));
```

**Review by ggerganov — APPROVED** [review]

**Comment by CISC** [comment]
Hey, it's worth running the server tests once in a while... :)
https://github.com/ggml-org/llama.cpp/actions/runs/26763150827/job/78882050608

**Comment by pwilkin** [comment]
Hey Claude please fix the tests so that they pass, make no mistakes!

**Comment by ServeurpersoCom** [comment]
I'm launching an agentic loop with CI/analysis.

**Comment by ServeurpersoCom** [comment]
Result :

Fix: server_n_outputs_max() ignored draft models that auto enable DRAFT_SIMPLE (empty speculative.types), under-reserving outputs and asserting on the verify batch; now keyed on has_dft().

```
diff --git a/tools/server/server-context.cpp b/tools/server/server-context.cpp
index b456d84b70..caee2e280d 100644
--- a/tools/server/server-context.cpp
+++ b/tools/server/server-context.cpp
@@ -75,6 +75,12 @@ static uint32_t server_n_outputs_max(const common_params & params) {
         }
     }

+    // draft model present without an explicit speculative type: init auto enables
+    // DRAFT_SIMPLE, so reserve room for the 1 + n_max tokens of the verify batch
+    if (params.speculative.has_dft()) {
+        n_outputs_per_seq = std::max<uint32_t>(n_outputs_per_seq, 1 + std::max(0, params.speculative.draft.n_max));
+    }
+
     const uint64_t n_outputs = (uint64_t) params.n_parallel * n_outputs_per_seq;

     return std::max<uint32_t>(1, std::min<uint64_t>(n_batch, n_outputs));
root@pod:/mnt/workspace/llama.cpp#
```

**Comment by ServeurpersoCom** [comment]
https://github.com/ggml-org/llama.cpp/pull/23987

**Comment by CMay** [comment]
This merge caused a performance regression from 90t/s to 60t/s on 7900XTX using the Vulkan backend.  VRAM usage did go down, though (thanks for that!).  Last time this happened, something was forcing sampling to run on the CPU rather than the GPU.  Is there anything in this commit that might cause that?  I'm not using any speculative features, just running Gemma 4 pretty vanilla with -b 2048 -ub 512.

**Comment by am17an** [comment]
@CMay I am not able to reproduce a performance regression with this change

**Comment by ggerganov** [comment]
@CMay I also don't reproduce on DGX Spark with Vulkan backend. Please open a separate issue with detailed logs (`-lv 4`).

**Comment by CMay** [comment]
It's not necessarily this commit's _fault_, but it started at this commit.

Took logs from both builds and fed them into Gemma 4.  It speculated that the differences in the compute buffer size between the 2 builds was almost certainly causing more CPU scheduling which surfaces any underlying CPU latency more.

It suggested increasing -ub even though it is only supposed to be relevant for prompt processing, because if the compute buffer is unified then it would also benefit token generation.

So I changed my -ub 512 to -ub 768 and bam, 90t/s.  Then tried specifying no -ub at all to let llama.cpp use its default, and it went back down to 60t/s.

**Edit: It turns out -kvu was enabled, but after disabling it token generation performance went back to normal regardless of the compute buffer size.  I'm not sure how that interacts with this commit to reduce token throughput or why changing -ub helped too, but I'm noting it here for anyone who reads this in the future.**

**Comment by am17an** [comment]
@CMay frankly, your comment makes no sense. It would be better if you just created an issue rather than feeding your logs into LLMs for a solution.
